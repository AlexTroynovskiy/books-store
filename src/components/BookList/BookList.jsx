import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import { Container, Typography, Switch, FormGroup, FormControlLabel, FormControl } from '@material-ui/core';

import { BookCard } from './components/BookCard';
import { Loading } from '../_shared/Loading';
import { ButtonComponent } from '../_shared/Button';
import { ErrorFetch } from '../_shared/ErrorFetch';

const useStyles = makeStyles({
  container: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    margin: '140px 0 100px 0',
    position: 'relative',
  },
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '20px',
    justifyContent: 'center',
  },
  form: {
    position: 'absolute',
    top: '-70px',
    left: '130px',
  },
});

export const BookList = ({
  books,
  requestBooks,
  errors,
  page,
  serverItemslength,
  getServerItemsLength,
  isSearching,
  changeSearchMode,
  search,
  activeCategories,
  getFavorites,
  favourites,
  isFavorite,
  setIsFavorite,
  favouritesBooks,
  removeFavourites,
}) => {
  const classes = useStyles();

  useEffect(() => {
    if (!books.length) {
      getServerItemsLength();
      requestBooks();
    }
  }, []);

  const handleClick = () => {
    requestBooks(page, search, activeCategories);
  };

  const showFavorites = () => {
    setIsFavorite(!isFavorite);
    if (favourites.length && !isFavorite) {
      getFavorites(favourites);
    } else if (isFavorite) {
      removeFavourites();
    }
  };

  const booksArr = isFavorite ? favouritesBooks : books;

  return (
    <Container component="div" className={classes.container}>
      <FormControl component="fieldset" className={classes.form}>
        <FormGroup aria-label="position" row>
          <FormControlLabel
            value="start"
            control={<Switch onChange={showFavorites} color="primary" />}
            label={isFavorite ? 'Show All' : 'Show Only Favourites'}
            labelPlacement="start"
          />
        </FormGroup>
      </FormControl>
      <Container component="div" className={classes.root}>
        {booksArr.length && !errors ? (
          <>
            {booksArr.map((book) => (
              <BookCard key={book.id} book={book} changeSearchMode={changeSearchMode} />
            ))}
          </>
        ) : isFavorite && !favourites.length ? (
          <Typography color="primary" variant="h3">
            You do not have any favourites!
          </Typography>
        ) : isSearching ? (
          <Typography variant="h3" color="primary">
            No results
          </Typography>
        ) : errors ? (
          <ErrorFetch />
        ) : (
          <Loading />
        )}
      </Container>
      {books.length && books.length < serverItemslength && !errors && !isFavorite ? (
        <ButtonComponent variant="contained" color="primary" onClick={handleClick}>
          Load more
        </ButtonComponent>
      ) : null}
    </Container>
  );
};

BookList.propTypes = {
  books: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      category: PropTypes.string.isRequired,
    })
  ).isRequired,
  favourites: PropTypes.arrayOf(PropTypes.number).isRequired,
  favouritesBooks: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      category: PropTypes.string.isRequired,
    })
  ).isRequired,
  requestBooks: PropTypes.func.isRequired,
  getServerItemsLength: PropTypes.func.isRequired,
  serverItemslength: PropTypes.number,
  page: PropTypes.number,
  errors: PropTypes.bool,
  isSearching: PropTypes.bool,
  isFavorite: PropTypes.bool,
  setIsFavorite: PropTypes.func.isRequired,
  changeSearchMode: PropTypes.func.isRequired,
  search: PropTypes.string,
  activeCategories: PropTypes.arrayOf(PropTypes.string).isRequired,
  getFavorites: PropTypes.func.isRequired,
  removeFavourites: PropTypes.func.isRequired,
};

BookList.defaultProps = {
  errors: false,
  serverItemslength: 24,
  page: 1,
  isSearching: false,
  isFavorite: false,
  search: '',
  favouritesBooks: [],
};
