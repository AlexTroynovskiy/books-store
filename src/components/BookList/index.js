import { connect } from 'react-redux';

import { BookList } from './BookList';
import {
  requestBooks,
  getServerItemsLength,
  changeSearchMode,
  getFavorites,
  removeFavourites,
} from '../../redux/Books/actions';

const mapStateToProps = (state) => {
  return {
    books: state.shop.books,
    errors: state.error.errors,
    page: state.shop.page,
    isSearching: state.shop.isSearching,
    serverItemslength: state.shop.serverItemslength,
    search: state.shop.search,
    activeCategories: state.shop.activeCategories,
    favourites: state.shop.favourites,
    favouritesBooks: state.shop.favouritesBooks,
  };
};

const mapDispatchToProps = {
  requestBooks,
  getServerItemsLength,
  changeSearchMode,
  getFavorites,
  removeFavourites,
};

export default connect(mapStateToProps, mapDispatchToProps)(BookList);
