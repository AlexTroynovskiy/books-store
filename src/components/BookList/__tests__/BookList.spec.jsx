import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import BookListConnect from '../';
import { BookList } from '../BookList';

describe('Component: BookList', () => {
  const initialState = {
    books: [],
    requestBooks: jest.fn(),
    errors: false,
    page: 1,
    serverItemslength: 0,
    getServerItemsLength: jest.fn(),
    isSearching: false,
    changeSearchMode: jest.fn(),
    search: '',
    activeCategories: [],
    getFavorites: jest.fn(),
    favourites: [],
    favouritesBooks: [],
    removeFavourites: jest.fn(),
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('Should be properly connected to the store', () => {
    const mockStore = configureStore();
    let store, wrapper;
    store = mockStore(initialState);

    beforeEach(() => {
      wrapper = shallow(
        <Provider store={store}>
          <BookListConnect {...initialState} />
        </Provider>
      );
    });

    test('To be in DOM', () => {
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('To have nested elements', () => {
    let wrapper;
    const props = { ...initialState, isFavorite: false, setIsFavorite: jest.fn() };

    test('To exist', () => {
      wrapper = shallow(<BookList {...props} />);
      expect(wrapper.exists()).toBe(true);
    });

    test('To match snapshot', () => {
      wrapper = renderer.create(<BookList {...props} />).toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
