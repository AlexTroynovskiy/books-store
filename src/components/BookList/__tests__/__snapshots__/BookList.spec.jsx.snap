// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: BookList To have nested elements To match snapshot 1`] = `
<div
  className="MuiContainer-root makeStyles-container-1 MuiContainer-maxWidthLg"
>
  <fieldset
    className="MuiFormControl-root makeStyles-form-3"
  >
    <div
      aria-label="position"
      className="MuiFormGroup-root MuiFormGroup-row"
    >
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementStart"
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-36 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                className="PrivateSwitchBase-input-39 MuiSwitch-input"
                disabled={false}
                onChange={[Function]}
                type="checkbox"
                value="start"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Show Only Favourites
        </span>
      </label>
    </div>
  </fieldset>
  <div
    className="MuiContainer-root makeStyles-root-2 MuiContainer-maxWidthLg"
  >
    <div
      className="makeStyles-root-82"
    >
      <div
        className="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
